File: src/games/would-you-rather/logic.ts
// src/games/would-you-rather/logic.ts — PURE logic. No clock/RNG/IO; seed via createInitialState.
import type { ColorToken, GameConfig, GameStateBase } from '../../sdk/types';
import { shuffle } from '../../engine/rng';
import { poolFor } from './content';
import { readOptions } from './config';
import type { WyrOptions } from './config';

export type WyrPhase = 'prompt' | 'collecting' | 'reveal' | 'results' | 'error';
export type Side = 'A' | 'B';
export type Majority = Side | 'tie';

export interface RoundRecord {
  itemId: string;
  countA: number;
  countB: number;
  majority: Majority;
  choices: Record<string, Side>;
}

export interface WyrState extends GameStateBase {
  phase: WyrPhase;
  options: WyrOptions;
  order: string[];
  index: number;
  total: number;
  playerIds: string[];
  playerNames: Record<string, string>;
  playerColors: Record<string, ColorToken | undefined>;
  choices: Record<string, Side>;
  quickCounts: { A: number; B: number } | null;
  handoffIndex: number;
  current: { countA: number; countB: number; majority: Majority } | null;
  history: RoundRecord[];
  scores: Record<string, number>;
  errorCode: 'EMPTY_DECK' | 'NOT_ENOUGH_PLAYERS' | null;
}

export type WyrAction =
  | { type: 'BEGIN_COLLECTION' }
  | { type: 'CHOOSE'; playerId: string; side: Side }
  | { type: 'UNDO_CHOICE'; playerId: string }
  | { type: 'ADVANCE_HANDOFF' }
  | { type: 'SET_QUICK_COUNTS'; A: number; B: number }
  | { type: 'REVEAL' }
  | { type: 'NEXT' }
  | { type: 'SKIP' };

function zeroScores(playerIds: string[]): Record<string, number> {
  const scores: Record<string, number> = {};
  playerIds.forEach((id) => (scores[id] = 0));
  return scores;
}

function tallyOf(s: WyrState): { countA: number; countB: number } {
  if (s.options.mode === 'quick') {
    return { countA: s.quickCounts?.A ?? 0, countB: s.quickCounts?.B ?? 0 };
  }
  let countA = 0;
  let countB = 0;
  Object.values(s.choices).forEach((side) => (side === 'A' ? countA++ : countB++));
  return { countA, countB };
}

const majorityOf = (countA: number, countB: number): Majority =>
  countA > countB ? 'A' : countB > countA ? 'B' : 'tie';

function applyRoundPoints(
  base: Record<string, number>,
  options: WyrOptions,
  choices: Record<string, Side>,
  majority: Majority,
): Record<string, number> {
  if (!options.awardMajorityPoints || options.mode === 'quick') return base;
  const scores = { ...base };
  for (const [pid, side] of Object.entries(choices)) {
    const win = majority === 'tie' ? options.tieCountsForBoth : side === majority;
    if (win) scores[pid] = (scores[pid] ?? 0) + 1;
  }
  return scores;
}

function recomputeScores(
  playerIds: string[],
  options: WyrOptions,
  history: RoundRecord[],
): Record<string, number> {
  let scores = zeroScores(playerIds);
  for (const r of history) scores = applyRoundPoints(scores, options, r.choices, r.majority);
  return scores;
}

export function createInitialState(config: GameConfig, seed: number): WyrState {
  const options = readOptions(config);
  const playerIds = config.players.map((p) => p.id as string);
  const playerNames: Record<string, string> = {};
  const playerColors: Record<string, ColorToken | undefined> = {};
  config.players.forEach((p) => {
    playerNames[p.id] = p.name;
    playerColors[p.id] = p.color;
  });

  const pool = shuffle(
    poolFor(options.deckId, options.maxIntensity).map((it) => it.id),
    seed,
  );
  const total = Math.min(options.roundLength, pool.length);
  const order = pool.slice(0, total);

  const errorCode: WyrState['errorCode'] =
    playerIds.length < 2 ? 'NOT_ENOUGH_PLAYERS' : order.length === 0 ? 'EMPTY_DECK' : null;

  return {
    v: 1,
    phase: errorCode ? 'error' : 'prompt',
    finished: false,
    options,
    order,
    index: 0,
    total,
    playerIds,
    playerNames,
    playerColors,
    choices: {},
    quickCounts: null,
    handoffIndex: 0,
    current: null,
    history: [],
    scores: zeroScores(playerIds),
    errorCode,
  };
}

function clearRound(): Pick<WyrState, 'choices' | 'quickCounts' | 'handoffIndex' | 'current'> {
  return { choices: {}, quickCounts: null, handoffIndex: 0, current: null };
}

export function reducer(state: WyrState, action: WyrAction): WyrState {
  const s = state;
  switch (action.type) {
    case 'BEGIN_COLLECTION': {
      if (s.phase !== 'prompt') return s;
      return { ...s, phase: 'collecting', choices: {}, quickCounts: null, handoffIndex: 0 };
    }
    case 'CHOOSE': {
      if (s.phase !== 'collecting' || s.options.mode !== 'vote') return s;
      if (!s.playerIds.includes(action.playerId)) return s;
      return { ...s, choices: { ...s.choices, [action.playerId]: action.side } };
    }
    case 'UNDO_CHOICE': {
      if (s.phase !== 'collecting') return s;
      if (!(action.playerId in s.choices)) return s;
      const choices = { ...s.choices };
      delete choices[action.playerId];
      return { ...s, choices };
    }
    case 'ADVANCE_HANDOFF': {
      if (s.phase !== 'collecting' || s.options.mode !== 'vote') return s;
      return { ...s, handoffIndex: Math.min(s.handoffIndex + 1, s.playerIds.length) };
    }
    case 'SET_QUICK_COUNTS': {
      if (s.phase !== 'collecting' || s.options.mode !== 'quick') return s;
      return { ...s, quickCounts: { A: Math.max(0, action.A), B: Math.max(0, action.B) } };
    }
    case 'REVEAL': {
      if (s.phase !== 'collecting') return s;
      const { countA, countB } = tallyOf(s);
      const majority = majorityOf(countA, countB);
      const scores = applyRoundPoints(s.scores, s.options, s.choices, majority);
      return { ...s, phase: 'reveal', current: { countA, countB, majority }, scores };
    }
    case 'NEXT': {
      if (s.phase !== 'reveal') return s;
      const cur = s.current ?? { countA: 0, countB: 0, majority: 'tie' as Majority };
      const record: RoundRecord = {
        itemId: s.order[s.index],
        countA: cur.countA,
        countB: cur.countB,
        majority: cur.majority,
        choices: { ...s.choices },
      };
      const history = [...s.history, record];
      const index = s.index + 1;
      const done = index >= s.total || index >= s.order.length;
      return {
        ...s,
        ...clearRound(),
        history,
        index,
        phase: done ? 'results' : 'prompt',
        finished: done,
      };
    }
    case 'SKIP': {
      if (s.phase !== 'prompt' && s.phase !== 'collecting' && s.phase !== 'reveal') return s;
      const index = s.index + 1;
      const done = index >= s.total || index >= s.order.length;
      const scores = recomputeScores(s.playerIds, s.options, s.history); // roll back this round
      return {
        ...s,
        ...clearRound(),
        index,
        scores,
        phase: done ? 'results' : 'prompt',
        finished: done,
      };
    }
    default:
      return s;
  }
}

/* ─────────────────────────  Pure selectors  ───────────────────────── */

export const currentItemId = (s: WyrState): string | null => s.order[s.index] ?? null;
export const currentVoterId = (s: WyrState): string | null =>
  s.options.mode === 'vote' && s.handoffIndex < s.playerIds.length
    ? s.playerIds[s.handoffIndex]
    : null;
export const everyoneVoted = (s: WyrState): boolean => s.handoffIndex >= s.playerIds.length;

export interface WyrStanding {
  id: string;
  score: number;
  rank: number;
}

export function standings(s: WyrState): WyrStanding[] {
  const order = new Map(s.playerIds.map((id, i) => [id, i]));
  const sorted = [...s.playerIds].sort(
    (a, b) => (s.scores[b] ?? 0) - (s.scores[a] ?? 0) || (order.get(a)! - order.get(b)!),
  );
  let rank = 0;
  let prev = NaN;
  return sorted.map((id, i) => {
    const sc = s.scores[id] ?? 0;
    if (sc !== prev) {
      rank = i + 1;
      prev = sc;
    }
    return { id, score: sc, rank };
  });
}

export function computeWinners(s: WyrState): string[] {
  if (!s.options.awardMajorityPoints || s.options.mode === 'quick') return [];
  return standings(s)
    .filter((r) => r.rank === 1 && r.score > 0)
    .map((r) => r.id);
}

export function sideWins(s: WyrState): { A: number; B: number } {
  let A = 0;
  let B = 0;
  s.history.forEach((r) => {
    if (r.majority === 'A') A++;
    else if (r.majority === 'B') B++;
  });
  return { A, B };
}

File: android-native/app/src/main/java/com/gamenight/party/game/would-you-rather/Logic.kt
package com.gamenight.party.game.wouldyourather

import com.gamenight.party.engine.Rng
import com.gamenight.party.model.ColorToken
import com.gamenight.party.model.Lang
import com.gamenight.party.model.LocalizedString
import com.gamenight.party.model.PlayerSeat

/**
 * PURE game logic for "Would You Rather" — a faithful Kotlin port of
 * src/games/would-you-rather/{config.ts,logic.ts}. No clock / RNG / IO except the [seed] threaded
 * into [createInitialState] via [Rng] (a fresh instance per call == the web's `shuffle(items, seed)`).
 * Side effects (sound/haptics) live in the Compose layer, not here.
 *
 * No-op transitions return the SAME state reference, mirroring the web reducer's `=== input` tests.
 */

// ──────────────────────────── Config / options ────────────────────────────

enum class WyrMode { VOTE, QUICK }

data class WyrOptions(
    val deckId: String = "classic",
    val maxIntensity: Intensity = Intensity.MEDIUM,
    val mode: WyrMode = WyrMode.VOTE,
    /** Number of prompts to play (clamped to the pool; a very large value plays the whole deck). */
    val roundLength: Int = 10,
    val awardMajorityPoints: Boolean = false,
    val tieCountsForBoth: Boolean = true,
)

val DEFAULT_OPTIONS: WyrOptions = WyrOptions()

/**
 * Mirrors config.ts `normalizeOptions`. The mode/intensity/booleans are already type-safe in Kotlin
 * (enums), and the deck is chosen from the valid set in Setup, so this only clamps [roundLength].
 */
fun normalizeOptions(o: WyrOptions): WyrOptions = o.copy(roundLength = o.roundLength.coerceIn(1, 999))

/**
 * The match configuration. Mirrors the webapp's `GameConfig` slice this game reads, plus the loaded
 * [content] so [createInitialState] can resolve the pool purely (the web reads a module-level pool).
 */
data class WyrConfig(
    val players: List<PlayerSeat>,
    val content: WyrContent,
    val lang: Lang = Lang.EN,
    val options: WyrOptions = DEFAULT_OPTIONS,
)

/** Mirrors config.ts `validateConfig`. Returns null when the config is playable. */
fun validateConfig(config: WyrConfig): List<LocalizedString>? {
    val o = normalizeOptions(config.options)
    val errors = mutableListOf<LocalizedString>()
    val n = config.players.size
    if (n < 2) errors += LocalizedString("Add at least 2 players", "حداقل ۲ بازیکن اضافه کن")
    if (n > 20) errors += LocalizedString("At most 20 players", "حداکثر ۲۰ بازیکن")
    if (config.content.poolFor(o.deckId, o.maxIntensity).isEmpty())
        errors += LocalizedString("No prompts at this intensity", "هیچ سوالی در این شدت نیست")
    return errors.ifEmpty { null }
}

// ──────────────────────────── State ────────────────────────────

enum class WyrPhase { PROMPT, COLLECTING, REVEAL, RESULTS, ERROR }

enum class Side { A, B }

enum class Majority { A, B, TIE }

enum class WyrError { EMPTY_DECK, NOT_ENOUGH_PLAYERS }

data class QuickCounts(val a: Int, val b: Int)

data class RoundCurrent(val countA: Int, val countB: Int, val majority: Majority)

data class RoundRecord(
    val itemId: String,
    val countA: Int,
    val countB: Int,
    val majority: Majority,
    val choices: Map<String, Side>,
)

data class WyrState(
    val v: Int,
    val phase: WyrPhase,
    val finished: Boolean,
    val options: WyrOptions,
    val order: List<String>,
    val index: Int,
    val total: Int,
    val playerIds: List<String>,
    val playerNames: Map<String, String>,
    val playerColors: Map<String, ColorToken?>,
    val choices: Map<String, Side>,
    val quickCounts: QuickCounts?,
    val handoffIndex: Int,
    val current: RoundCurrent?,
    val history: List<RoundRecord>,
    val scores: Map<String, Int>,
    val errorCode: WyrError?,
)

// ──────────────────────────── Actions ────────────────────────────

sealed interface WyrAction {
    data object BeginCollection : WyrAction
    data class Choose(val playerId: String, val side: Side) : WyrAction
    data class UndoChoice(val playerId: String) : WyrAction
    data object AdvanceHandoff : WyrAction
    data class SetQuickCounts(val a: Int, val b: Int) : WyrAction
    data object Reveal : WyrAction
    data object Next : WyrAction
    data object Skip : WyrAction
    data object EndGame : WyrAction
}

// ──────────────────────────── Scoring helpers ────────────────────────────

private fun zeroScores(playerIds: List<String>): Map<String, Int> = playerIds.associateWith { 0 }

private fun tallyOf(s: WyrState): Pair<Int, Int> {
    if (s.options.mode == WyrMode.QUICK) {
        return Pair(s.quickCounts?.a ?: 0, s.quickCounts?.b ?: 0)
    }
    var countA = 0
    var countB = 0
    s.choices.values.forEach { if (it == Side.A) countA++ else countB++ }
    return Pair(countA, countB)
}

private fun majorityOf(countA: Int, countB: Int): Majority =
    if (countA > countB) Majority.A else if (countB > countA) Majority.B else Majority.TIE

private fun sideMatchesMajority(side: Side, majority: Majority): Boolean =
    (side == Side.A && majority == Majority.A) || (side == Side.B && majority == Majority.B)

private fun applyRoundPoints(
    base: Map<String, Int>,
    options: WyrOptions,
    choices: Map<String, Side>,
    majority: Majority,
): Map<String, Int> {
    if (!options.awardMajorityPoints || options.mode == WyrMode.QUICK) return base
    val scores = LinkedHashMap(base)
    for ((pid, side) in choices) {
        val win = if (majority == Majority.TIE) options.tieCountsForBoth else sideMatchesMajority(side, majority)
        if (win) scores[pid] = (scores[pid] ?: 0) + 1
    }
    return scores
}

private fun recomputeScores(
    playerIds: List<String>,
    options: WyrOptions,
    history: List<RoundRecord>,
): Map<String, Int> {
    var scores = zeroScores(playerIds)
    for (r in history) scores = applyRoundPoints(scores, options, r.choices, r.majority)
    return scores
}

// ──────────────────────────── Initial state ────────────────────────────

fun createInitialState(config: WyrConfig, seed: Int): WyrState {
    val options = normalizeOptions(config.options)
    val playerIds = config.players.map { it.id }
    val playerNames: Map<String, String> = config.players.associate { it.id to it.name }
    val playerColors: Map<String, ColorToken?> = config.players.associate { it.id to it.color }

    // Fresh Rng per call mirrors the web's `shuffle(poolFor(...).map(id), seed)`.
    val pool = Rng(seed).shuffle(config.content.poolFor(options.deckId, options.maxIntensity).map { it.id })
    val total = minOf(options.roundLength, pool.size)
    val order = pool.take(total)

    val errorCode: WyrError? = when {
        playerIds.size < 2 -> WyrError.NOT_ENOUGH_PLAYERS
        order.isEmpty() -> WyrError.EMPTY_DECK
        else -> null
    }

    return WyrState(
        v = 1,
        phase = if (errorCode != null) WyrPhase.ERROR else WyrPhase.PROMPT,
        finished = false,
        options = options,
        order = order,
        index = 0,
        total = total,
        playerIds = playerIds,
        playerNames = playerNames,
        playerColors = playerColors,
        choices = emptyMap(),
        quickCounts = null,
        handoffIndex = 0,
        current = null,
        history = emptyList(),
        scores = zeroScores(playerIds),
        errorCode = errorCode,
    )
}

// ──────────────────────────── Reducer ────────────────────────────

fun reducer(state: WyrState, action: WyrAction): WyrState {
    val s = state
    return when (action) {
        is WyrAction.BeginCollection -> {
            if (s.phase != WyrPhase.PROMPT) return s
            s.copy(phase = WyrPhase.COLLECTING, choices = emptyMap(), quickCounts = null, handoffIndex = 0)
        }

        is WyrAction.Choose -> {
            if (s.phase != WyrPhase.COLLECTING || s.options.mode != WyrMode.VOTE) return s
            if (action.playerId !in s.playerIds) return s
            s.copy(choices = s.choices + (action.playerId to action.side))
        }

        is WyrAction.UndoChoice -> {
            if (s.phase != WyrPhase.COLLECTING) return s
            if (action.playerId !in s.choices) return s
            s.copy(choices = s.choices - action.playerId)
        }

        is WyrAction.AdvanceHandoff -> {
            if (s.phase != WyrPhase.COLLECTING || s.options.mode != WyrMode.VOTE) return s
            s.copy(handoffIndex = minOf(s.handoffIndex + 1, s.playerIds.size))
        }

        is WyrAction.SetQuickCounts -> {
            if (s.phase != WyrPhase.COLLECTING || s.options.mode != WyrMode.QUICK) return s
            s.copy(quickCounts = QuickCounts(maxOf(0, action.a), maxOf(0, action.b)))
        }

        is WyrAction.Reveal -> {
            if (s.phase != WyrPhase.COLLECTING) return s
            val (countA, countB) = tallyOf(s)
            val majority = majorityOf(countA, countB)
            val scores = applyRoundPoints(s.scores, s.options, s.choices, majority)
            s.copy(phase = WyrPhase.REVEAL, current = RoundCurrent(countA, countB, majority), scores = scores)
        }

        is WyrAction.Next -> {
            if (s.phase != WyrPhase.REVEAL) return s
            val cur = s.current ?: RoundCurrent(0, 0, Majority.TIE)
            val record = RoundRecord(
                itemId = s.order.getOrElse(s.index) { "" },
                countA = cur.countA,
                countB = cur.countB,
                majority = cur.majority,
                choices = s.choices,
            )
            val history = s.history + record
            val index = s.index + 1
            val done = index >= s.total || index >= s.order.size
            s.copy(
                choices = emptyMap(),
                quickCounts = null,
                handoffIndex = 0,
                current = null,
                history = history,
                index = index,
                phase = if (done) WyrPhase.RESULTS else WyrPhase.PROMPT,
                finished = done,
            )
        }

        is WyrAction.Skip -> {
            if (s.phase != WyrPhase.PROMPT && s.phase != WyrPhase.COLLECTING && s.phase != WyrPhase.REVEAL) return s
            val index = s.index + 1
            val done = index >= s.total || index >= s.order.size
            // Roll back any points the (now-discarded) current round may have applied.
            val scores = recomputeScores(s.playerIds, s.options, s.history)
            s.copy(
                choices = emptyMap(),
                quickCounts = null,
                handoffIndex = 0,
                current = null,
                index = index,
                scores = scores,
                phase = if (done) WyrPhase.RESULTS else WyrPhase.PROMPT,
                finished = done,
            )
        }

        is WyrAction.EndGame -> {
            if (s.finished) return s
            // End the match now and show standings so far. Finalize scores from the COMPLETED rounds
            // only (mirrors Skip's rollback of the in-progress round) so the Results view — which reads
            // scores for standings/winners and history for side wins — stays internally consistent.
            val scores = recomputeScores(s.playerIds, s.options, s.history)
            s.copy(
                choices = emptyMap(),
                quickCounts = null,
                handoffIndex = 0,
                current = null,
                phase = WyrPhase.RESULTS,
                scores = scores,
                finished = true,
            )
        }
    }
}

// ──────────────────────────── Pure selectors ────────────────────────────

fun currentItemId(s: WyrState): String? = s.order.getOrNull(s.index)

fun currentVoterId(s: WyrState): String? =
    if (s.options.mode == WyrMode.VOTE && s.handoffIndex < s.playerIds.size) s.playerIds[s.handoffIndex] else null

fun everyoneVoted(s: WyrState): Boolean = s.handoffIndex >= s.playerIds.size

data class WyrStanding(val id: String, val score: Int, val rank: Int)

fun standings(s: WyrState): List<WyrStanding> {
    val orderIdx = s.playerIds.withIndex().associate { (i, id) -> id to i }
    val sorted = s.playerIds.sortedWith(
        compareByDescending<String> { s.scores[it] ?: 0 }.thenBy { orderIdx[it] ?: 0 },
    )
    var rank = 0
    var prev = 0
    var hasPrev = false
    return sorted.mapIndexed { i, id ->
        val sc = s.scores[id] ?: 0
        if (!hasPrev || sc != prev) {
            rank = i + 1
            prev = sc
            hasPrev = true
        }
        WyrStanding(id, sc, rank)
    }
}

fun computeWinners(s: WyrState): List<String> {
    if (!s.options.awardMajorityPoints || s.options.mode == WyrMode.QUICK) return emptyList()
    return standings(s).filter { it.rank == 1 && it.score > 0 }.map { it.id }
}

/** Total rounds each side won across [WyrState.history]. */
fun sideWins(s: WyrState): Pair<Int, Int> {
    var a = 0
    var b = 0
    s.history.forEach { if (it.majority == Majority.A) a++ else if (it.majority == Majority.B) b++ }
    return Pair(a, b)
}
